src/client/debugger/jupyter/debuggingManager.ts

```diff
@@ -0,0 +1,213 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+'use strict';
+
+import { inject, injectable } from 'inversify';
+import * as vscode from 'vscode';
+import * as path from 'path';
+import { IKernelProvider } from '../../datascience/jupyter/kernels/types';
+import { IDisposable } from '../../common/types';
+import { KernelDebugAdapter } from './kernelDebugAdapter';
+import { IDebuggingCellMap, INotebookProvider } from '../../datascience/types';
+import { IExtensionSingleActivationService } from '../../activation/types';
+import { ServerStatus } from '../../../datascience-ui/interactive-common/mainState';
+import { INotebookControllerManager } from '../../datascience/notebook/types';
+
+class Debugger {
+    private resolveFunc?: (value: vscode.DebugSession) => void;
+    private rejectFunc?: (reason?: Error) => void;
+
+    readonly session: Promise<vscode.DebugSession>;
+
+    constructor(public readonly document: vscode.NotebookDocument) {
+        this.session = new Promise<vscode.DebugSession>((resolve, reject) => {
+            this.resolveFunc = resolve;
+            this.rejectFunc = reject;
+
+            vscode.debug
+                .startDebugging(undefined, {
+                    type: 'kernel',
+                    name: `${path.basename(document.uri.toString())}`,
+                    request: 'attach',
+                    internalConsoleOptions: 'neverOpen',
+                    __document: document.uri.toString()
+                })
+                .then(undefined, reject);
+        });
+    }
+
+    resolve(session: vscode.DebugSession) {
+        if (this.resolveFunc) {
+            this.resolveFunc(session);
+        }
+    }
+
+    reject(reason: Error) {
+        if (this.rejectFunc) {
+            this.rejectFunc(reason);
+        }
+    }
+
+    async stop() {
+        void vscode.debug.stopDebugging(await this.session);
+    }
+}
+
+/**
+ * The DebuggingManager maintains the mapping between notebook documents and debug sessions.
+ */
+@injectable()
+export class DebuggingManager implements IExtensionSingleActivationService {
+    private notebookToDebugger = new Map<vscode.NotebookDocument, Debugger>();
+    private readonly disposables: IDisposable[] = [];
+
+    public constructor(
+        @inject(IKernelProvider) private kernelProvider: IKernelProvider,
+        @inject(INotebookProvider) private notebookProvider: INotebookProvider,
+        @inject(IDebuggingCellMap) private debuggingCellMap: IDebuggingCellMap,
+        @inject(INotebookControllerManager) private readonly notebookControllerManager: INotebookControllerManager
+    ) {}
+
+    public async activate() {
+        vscode.debug.breakpoints; // start to fetch breakpoints
+
+        this.disposables.push(
```

Comment:
I don't see the dispose pattern or interface being implemented to go along with this. Probably just IDisposableRegistry.